### PR TITLE
new release ocaml-freestanding.0.7.0

### PR DESCRIPTION
**Pull request by dinosaure**
## v0.7.0 (2022-01-03)

* **NOTE**: This release is a part of the MirageOS 4.0 release and, at this stage, `ocaml-freestanding.0.7.0` will **not** continue to support MirageOS 3 unikernels. `ocaml-freestanding` becomes a real cross-compiler used then by `dune` (with the cross-compilation option) to compile libraries including C stubs. For the MirageOS 4.0 perspective, libraries which include C stubs does not need anymore to compile multiple times C artifacts depending on what `ocaml-freestanding` provides. The `dune` context will help the compilation of these C stubs according the target chosen by the end-user (with the `mirage` tool).

* Add compatibility with the solo5 0.7.0 package split. (mirage/ocaml-freestanding#104)
* Build a freestanding cross-compiler to use with the ocamlfind toolchain feature. This cross-compiler is able to build partial executables to link with a solo5 bindings library. (mirage/ocaml-freestanding#104)
* Fix the OpenBSD 7 support (mirage/ocaml-freestanding#104)
* Remove `pkg-config` (mirage/ocaml-freestanding#104)

### CODE REVIEW

**Comment by dinosaure** [comment]
We probably need to add constraints on:
- `gmp-freestanding`
- `zarith-freestanding`

/cc @TheLortex, @hannesm

**Comment by TheLortex** [comment]
What about the other revdeps, such as `bigstringaf`, `checkseum`, `digestif` ? They should also get the constraints

**Comment by dinosaure** [comment]
> What about the other revdeps, such as bigstringaf, checkseum, digestif ? They should also get the constraints

These packages are more smart than `zarith-freestanding`/`gmp-freestanding`. If `pkg-config` does not exists, they just `touch libfreestanding_*.a` and do an usual installation. Then, put a constraint on them will require a release of them then without their C artifacts (which is much more work). I think the situation is good enough for them - and we will just waiting another release of them to delete their `freestanding` supports.

**Comment by hannesm** [comment]
yes, ti me this looks good -- including upper bounds on gmp-freestanding and zarith-freestanding.

Maybe worth to investigate the extras:
```
flambda-fp

    ocaml-freestanding-cross-aarch64.0.7.0 (failed: frame pointers not supported on this platform)
    ocaml-freestanding.0.7.0 (failed: frame pointers not supported on this platform)

fp

    ocaml-freestanding-cross-aarch64.0.7.0 (failed: frame pointers not supported on this platform)
    ocaml-freestanding.0.7.0 (failed: frame pointers not supported on this platform)

nnpchecker

    ocaml-freestanding-cross-aarch64.0.7.0 (failed: naked pointers checker not supported on this platform)
    ocaml-freestanding.0.7.0 (failed: naked pointers checker not supported on this platform)
```

**Comment by dinosaure** [comment]
`solo5-bindings-*` requires at least GCC 4.0 but I don't know how to check that - if we have a `conf-gcc` package or something like that. Then, as @hannesm said, `flambda-fp`, `fp` and `nnpchecker` seems accurate errors about this release - others errors seems unrelated.

**Comment by TheLortex** [comment]
> It seems however that we have a problem on Archlinux:

It's a bit strange because it's an exception that comes from `obuilder`

**Comment by dinosaure** [comment]
The current test to check if we handle frame-pointer is:
```sh
  case "$host,$CC" in #(
  x86_64-*-linux*,gcc*|x86_64-*-linux*,clang*) :
    common_cflags="$common_cflags -g  -fno-omit-frame-pointer"
      frame_pointers=true
      $as_echo "#define WITH_FRAME_POINTERS 1" >>confdefs.h

      { $as_echo "$as_me:${as_lineno-$LINENO}: using frame pointers" >&5
$as_echo "$as_me: using frame pointers" >&6;} ;; #(
  *) :
    as_fn_error $? "frame pointers not supported on this platform" "$LINENO" 5

```

Which obviously fails because the compiler is: `x86_64-solo5-none-static`. Unfortunately, we can not lie on this side - we must keep our toolchain. The best will be to say that `ocaml-freestanding` is not allowed with a compiler with frame-pointers/naked-pointers.

**Comment by hannesm** [comment]
what we could do is to add `-fno-omit-frame-pointer` to the cflags if desired. but until anyone requests this feature, we can also just remove the `fp` (see https://github.com/hannesm/ocaml-freestanding/commit/c47deb49bba0cb59b4fbe2b105bdb359ecda8f4e). the afl option isn't worth it, since with freestanding we're not supporting the C runtime (runtime/afl.c from OCaml) part for afl communication.

File: packages/ocaml-freestanding-cross-aarch64/ocaml-freestanding-cross-aarch64.0.7.0/opam [hunk 1]
```diff
@@ -33,6 +33,9 @@ conflicts: [
   "solo5-kernel-ukvm"
   "solo5-kernel-virtio"
   "solo5-kernel-muen"
+  "ocaml-option-fp"
+  "ocaml-option-nnp"
```
Comment (hannesm) [inline packages/ocaml-freestanding-cross-aarch64/ocaml-freestanding-cross-aarch64.0.7.0/opam:37]: the nnp option is fine afaict. the nnpchecker is not fine.

**Comment by dinosaure** [comment]
This PR seems ready to merge. About `zarith-freestanding`/`gmp-freestanding`:
1) we don't use them anymore for MirageOS 4.0
2) I don't know how to constraint them (and specially `solo5-bindings-*` packages) to require, at least, `gcc.4.*`

**Comment by hannesm** [comment]
I looked through the failures, and indeed they are unrelated -- some sandbox issues (9p gets EROFS), some other issues in packages -- but nothing related to the new ocaml-freestanding. I fully support a merge of this PR.

**Pull request merged**